File: 0994-rotting-oranges/0994-rotting-oranges.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        // for each rotten fruit, start bfs and update the nearest fresh fruit time, 
        // return the max minimum time of that fresh fruit turn to rotten fruit
        int minimumElapse = -1;
        queue<vector<int>> fruits;
        vector<vector<bool>> visited(grid.size(), vector<bool>(grid[0].size(), false));
        vector<pair<int, int>> directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        int numFreshFruits = 0;
        int numVisitedFreshFruits = 0;

        for (auto i = 0; i < grid.size(); ++i) {
            for (auto j = 0; j < grid[0].size(); ++j) {
                if (grid[i][j] == 2) {
                    vector<int> fruit = {i, j, 0};
                    fruits.push(fruit);
                } else if (grid[i][j] == 1) {
                    numFreshFruits++;
                }
            }
        }

        if (numFreshFruits == 0) {
            return 0;
        }

        while (!fruits.empty()) {
            auto size = fruits.size();

            for (auto i = 0; i < size; ++i) {
                auto fruit = fruits.front();
                fruits.pop();

                if (visited[fruit[0]][fruit[1]]) {
                    continue;
                }
                visited[fruit[0]][fruit[1]] = true;

                if (grid[fruit[0]][fruit[1]] == 1) {
                    numVisitedFreshFruits++;
                    minimumElapse = max(minimumElapse, fruit[2]);
                }

                for (const auto& [dx, dy]: directions) {
                    int nx = fruit[0] + dx;
                    int ny = fruit[1] + dy;

                    if (nx >= 0 && nx < grid.size() &&
                        ny >= 0 && ny < grid[0].size() &&
                        !visited[nx][ny] && grid[nx][ny] == 1) {
                            vector<int> neighborFruit = {nx, ny, fruit[2] + 1};
                            fruits.push(neighborFruit);
                        }
                }
            }   
        }
        return numVisitedFreshFruits < numFreshFruits ? -1 : minimumElapse;
    }
};

File: 0994-rotting-oranges/0994-rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])
        # time complexity o(nm), calcuate roten and vvisited and fresh num  + BFS = o(V+E) = o(nm)

        fresh_num = sum([1 for i in range(m) for j in range(n) if grid[i][j] == 1])

        minimum_time = 0
        visited = [[False for j in range(n)] for i in range(m)]

        rottens = [(i, j, 0) for i in range(m) for j in range(n) if grid[i][j] == 2]
        queue = deque(rottens)

        while queue:
            x, y, time = queue.popleft()

            minimum_time = max(minimum_time, time)
            
            for dx, dy in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
                nx = x + dx
                ny = y + dy

                if m > nx >= 0 and n > ny >= 0 and grid[nx][ny] == 1 and not visited[nx][ny]:
                    fresh_num -= 1
                    visited[nx][ny] = True
                    queue.append((nx, ny, time + 1))
        
        return minimum_time if fresh_num == 0 else -1